README.md
# (MODX)EvolutionCMS.plugins.ManagerManager.mm_hideEmpty

A widget for ManagerManager plugin that allows to hide all empty sections and tabs.


## Requires

* PHP >= 5.4
* [(MODX)EvolutionCMS.plugins.ManagerManager](https://code.divandesign.biz/modx/managermanager) >= 0.7


## Installation

To install you must unzip the archive to `/assets/plungins/managermanager/widgets/mm_hideempty/`.


You may also read this documentation:
* [(MODX)EvolutionCMS.plugins.ManagerManager](https://code.divandesign.biz/modx/managermanager).
* [(MODX)EvolutionCMS.modules.ddMMEditor](https://code.divandesign.biz/modx/ddmmeditor).


## Parameters description

* `$params`
	* Desctription: Parameters, the pass-by-name style is used.
	* Valid values:
		* `stdClass`
		* `arrayAssociative`
	* **Required**
	
* `$params->roles`
	* Desctription: The CMS user roles that the widget is applied to.
	* Valid values:
		* `array`
		* `stringCommaSeparated`
		* `''` — when this parameter is empty then widget is applied to the all roles
	* Default value: `''`
	
* `$params->roles[$i]`
	* Desctription: CMS user role ID.
	* Valid values: `integer`
	* **Required**
	
* `$params->templates`
	* Desctription: Document templates IDs for which the widget is applied to.
	* Valid values:
		* `array`
		* `stringCommaSeparated`
		* `''` — empty value means the widget is applying to all templates
	* Default value: `''`
	
* `$params->templates[$i]`
	* Desctription: Template ID.
	* Valid values: `integer`
	* **Required**


## CMS events

* `OnDocFormPrerender`
* `OnDocFormRender`


## Examples


### Apply for all roles and templates

```php
mm_hideEmpty();
```


## Links

* [Home page](https://code.divandesign.biz/modx/mm_hideempty)
* [Telegram chat](https://t.me/dd_code)
* [Packagist](https://packagist.org/packages/dd/evolutioncms-plugins-managermanager-mm_hideempty)
* [GitHub](https://github.com/DivanDesign/EvolutionCMS.plugins.ManagerManager.mm_hideEmpty)


<link rel="stylesheet" type="text/css" href="https://raw.githack.com/DivanDesign/CSS.ddMarkdown/master/style.min.css" />

CHANGELOG.md
# (MODX)EvolutionCMS.plugins.ManagerManager.mm_hideEmpty changelog


## Version 1.0 (2022-05-23)
* \+ Sections and tabs containing only hidden children are also considered as empty.
* \+ Some native section bodies have no the `id` attribute (e. g. TVs). Now the widget considers these sections too.
* \+ Composer.json.
* \+ README.
* \+ README_ru.
* \+ CHANGELOG.
* \+ CHANGELOG_ru.


## Version 0.2 (2016-11-05)
* \* Attention! PHP >= 5.4 is required.
* \* Attention! (MODX)Evolution.plugins.ManagerManager >= 0.7 is required.
* \* The widget now uses named parameters (with backward compatibility).
* \* JS has been moved out to a standalone file.


## Version 0.1 (2016-04-25)
* \+ The first release.


<link rel="stylesheet" type="text/css" href="https://raw.githack.com/DivanDesign/CSS.ddMarkdown/master/style.min.css" />
<style>ul{list-style:none;}</style>

jQuery.ddMM.mm_hideEmpty.js
/**
 * jQuery.ddMM.mm_hideEmpty
 * @version 1.1 (2022-05-23)
 * 
 * @copyright 2016–2022
 */

(function($){
$.ddMM.mm_hideEmpty = {
	/**
	 * @method hideEmptySections
	 * @version 1.0.3 (2022-05-23)
	 * 
	 * @desc Hide empty sections.
	 * 
	 * @returns {void}
	 */
	hideEmptySections: function(){
		this
			.getEmptyElements('.sectionBody[id]')
			.each(function(){
				var
					$this = $(this),
					id =
						$this
							.attr('id')
							.match(/(.+)_[^_]+$/)
							[1]
				;
				
				//Section header
				$('#' + id + '_header').hide();
				
				//Section body
				$this.hide();
			})
		;
	},
	
	/**
	 * @method hideEmptyTabs
	 * @version 1.0.3 (2022-05-23)
	 * 
	 * @desc Hide empty tabs.
	 * 
	 * @returns {void}
	 */
	hideEmptyTabs: function(){
		this
			.getEmptyElements('.tab-pane .tab-page')
			.each(function(){
				var $this = $(this);
				
				//Page
				$this.hide();
				
				//Navigation item
				$('.tab-pane .tab-row .tab')
					.eq(
						$this
							.get(0)
							.tabPage
							.index
					)
					.hide()
				;
			})
		;
	},
	
	/**
	 * @method getEmptyElements
	 * @version 1.1 (2022-05-23)
	 * 
	 * @desc Get parents that has no visible children.
	 * 
	 * @param selector {string} — jQuery selector to get elements.
	 * 
	 * @returns {jQuery}
	 */
	getEmptyElements: function(selector){
		var
			$elements = $(selector),
			$elementsVisible = $elements.filter(':visible'),
			$elementsInvisible = $elements.not($elementsVisible)
		;
		
		return (
			//First, get empty visible elements
			$elementsVisible
				//That has no visible inputs
				.not(':has([name]:visible)')
				//And has no visible iframes
				.not(':has(iframe:visible)')
				
				//Second, get empty invisible elements
				.add(
					$elementsInvisible
						//That has no inputs
						.not(':has([name])')
						//And has no iframes
						.not(':has(iframe)')
				)
		);
	}
};
})(jQuery);

mm_hideempty.php
<?php
/**
 * mm_hideEmpty
 * @version 1.0 (2022-05-23)
 * 
 * @see README.md
 * 
 * @link http://code.divandesign.ru/modx/mm_hideempty
 * @link https://github.com/MrSwed/MODXEvo.plugin.ManagerManager.mm_hideEmpty
 * 
 * @copyright 2016–2022
 */

function mm_hideEmpty($params = []){
	//For backward compatibility
	if (
		!is_array($params) &&
		!is_object($params)
	){
		//Convert ordered list of params to named
		$params = \ddTools::orderedParamsToNamed([
			'paramsList' => func_get_args(),
			'compliance' => [
				'roles',
				'templates'
			]
		]);
	}
	
	$params = \DDTools\ObjectTools::extend([
		'objects' => [
			//Defaults
			(object) [
				'roles' => '',
				'templates' => ''
			],
			$params
		]
	]);
	
	if (
		!useThisRule(
			$params->roles,
			$params->templates
		)
	){
		return;
	}
	
	global $modx;
	
	if ($modx->Event->name == 'OnDocFormPrerender'){
		//The main js file including
		$output = includeJsCss(
			$modx->getConfig('site_url') . 'assets/plugins/managermanager/widgets/mm_hideempty/jQuery.ddMM.mm_hideEmpty.js',
			'html',
			'jQuery.ddMM.mm_hideEmpty',
			'1.2.1'
		);
		
		$modx->Event->output($output);
	}elseif ($modx->Event->name == 'OnDocFormRender'){
		$output = '//---------- mm_hideEmpty :: Begin -----' . PHP_EOL;
		
		$output .= '$j.ddMM.mm_hideEmpty.hideEmptySections();' . PHP_EOL;
		$output .= '$j.ddMM.mm_hideEmpty.hideEmptyTabs();' . PHP_EOL;
		
		$output .= '//---------- mm_hideEmpty :: End -----' . PHP_EOL;
		
		$modx->Event->output($output);
	}
}
?>